Update todomvc-aurelia 0.2.0 (using Aurelia 1.0.0.beta-1)

mhoyer opened the pull request:
- Fixed bugs
  - #1492 - incorrect ref path in Firefox
  - #1543 - could not be reproduced for me with todomvc-aurelia 0.2.0
- Updated to latest JSPM/NPM dependencies
  - Aurelia 1.0.0-beta.1
  - TodoMVC assets (incl. updated views)

Further information can be found in changelog of original todomvc-aurelia repo:
https://github.com/mhoyer/todomvc-aurelia/blob/release-0.2/README.md#changelog

mhoyer commented:
Help wanted: 
- Is there a general approach on how to provide Todo apps to the master:examples that are based on transpiled sources?
  - Should I ship the original sources only - or besides transpiled ones?
  - Or is it more appreciated to reference just to original git repository where the whole project has a build chain and some kind of logical project structure?
- How should I ever pass the `jscs` linting rules when relying on SystemJS or other 3rd party dependencies?

samccone commented:
Hi @mhoyer funny you ask!

YES there is (as of a few days ago)
https://github.com/tastejs/todomvc/issues/1541

I would love to work with you in implementing the first "perf" ready example (we can do it in another PR if that is ok?)

samccone commented:
> Should I ship the original sources only - or besides transpiled ones?

Ship both!

> Or is it more appreciated to reference just to original git repository where the whole project has a build chain and some kind of logical project structure?

see 1541 ;)


For jscs @mhoyer please add the vendor files and files that should not be linted to our root .jscsrc file,
our test suite is very extensive and you can mostly trust its state. The next step for the PR is to get this all green.

Thanks for the work already!

samccone commented:
@mhoyer let me know if you need any help or whatever

mhoyer commented:
@samccone I pushed some new attempts to the PR, but it seems there are still some issues. Here is my walkthrough of fixes:

0. The given JSCS@1.13.1 seems to not support ES6. Thus, I updated it in ```./tooling``` to latest 2.7.0
0. After that, JSCS had an issue with ```excludeFiles``` option set in ```.jscsrc``` when running ```jscs``` in a subfolder. See https://github.com/jscs-dev/node-jscs/issues/669. I fixed that too by updating the *lint* npm script in ```./tooling/package.json``` to cd up to root before running the linter.
0. Now linting passes, but the tests now break my PR. :-(

Regarding the test suite I was already wonding with my [initial PR for Aurelia](https://github.com/tastejs/todomvc/pull/1163#issue-57694998) (first bullet point of § *Cons* section). IMHO the given suite is not perfectly fitting for those SPA frameworks that build up the UI dynamically. It would be interesting to see which examples are actually passing this suite. ;-)

Off-topic: is there any plan to switch to a "better" webdriver framework like [webdriver.io](https://www.npmjs.com/package/webdriverio)? IMHO it feels to be more maintained and documented compared to the original SeleniumHQ [selenium-webdriver](https://www.npmjs.com/package/selenium-webdriver).

samccone commented:
1. awesome!!!
2. hurray  :sparkles: 

(for 1 and 2 how would you feel about PRing those back into master independent of this PR)

3. Almost all of the main page implementations (and a majority of the secondary) pass on CI 100% and locally

For instance

https://github.com/tastejs/todomvc/pull/1538
https://github.com/tastejs/todomvc/pull/1526

(just look at the closed/merged PRs ... you can see that things do indeed work ... and the integration suite is actually really reliable)

If you feel like the tests are failing incorrectly... we can improve the suite to better handle "async" dom building... I have dropped a significant amount of time into making sure the suite is rock solid, so I would like to see it be reliable

https://github.com/tastejs/todomvc/pull/1371
https://github.com/tastejs/todomvc/pull/1499

> Off-topic: is there any plan to switch to a "better" webdriver framework like webdriver.io? IMHO it feels to be more maintained and documented compared to the original SeleniumHQ selenium-webdriver.

Heh, I have had the opposite experience with webdriver.io (@wkf can also vouch for this) in 2014 however.... if you want to take a stab... I am not 100% opposed 

---------

As always I realllllly appreciate this work and your thoroughness! :goat:

mhoyer commented:
Hi @samccone,

### JSCS update

> 0. awesome!!!
> 0. hurray  :sparkles:

> (for 1 and 2 how would you feel about PRing those back into master independent of this PR)

So I did as you surely noticed already. ;-)

### Broken unit tests

I think we have a common problem here. With latest [todomvc-app-template/index.html](https://github.com/tastejs/todomvc-app-template/blob/master/index.html) template and [CSS for todomvc](https://github.com/tastejs/todomvc-app-css/commit/a722ad8119d559228cba1e4895dabacffffcdb2b) a breaking change was introduced by replacing all id attributes with CSS classes. This is one major reason why most of the tests fail in my Aurelia version where I [already updated everything to this new approach](https://github.com/mhoyer/todomvc-aurelia/commit/dc4ff2ced73d0e04d62c6cdb1d0aaa004c6c45ca).

It can already be seen for the very first test that fails:

```
$ npm test -- "--framework=aurelia"

> todomvc-browser-tests@ test d:\projects\todomvc\tests
> mocha allTests.js --no-timeouts --reporter spec "--framework=aurelia"

  TodoMVC - aurelia
    When page is initially opened
      1) should focus on the todo input field

  0 passing (3s)
  1 failing

  1) TodoMVC - aurelia When page is initially opened should focus on the todo input field:

      AssertionError: The focused element did not have the expected id new-todo
      + expected - actual
```

**Q:** How should we now continue? Should I adjust the tests to pass while relying on the latest CSS and index.html template?

mhoyer commented:
Ah sorry, I see this case of supporting ID and CSS seems to already be covered with [```ensureAppIsVisible()```](https://github.com/tastejs/todomvc/blob/master/tests/page.js#L207).

samccone commented:
yep that is the function that flips the bit to use classes or ids

So the next question is, why does the first test fail? Why is it unable to find the element?

mhoyer commented:
Well the first test itself seems a bit strange to me:

```
// from test.js
test.describe('When page is initially opened', function () {
	test.it('should focus on the todo input field', function () {
		testOps.assertFocussedElementId('new-todo');
	});
});

// from testOperations.js
this.assertFocussedElementId = function (expectedId) {
	page.getFocussedTagName().then(function (id) {
		assert.notEqual(id.indexOf(expectedId), -1, 'The focused element did not have the expected id ' + expectedId);
	});
};

// from page.js
this.getFocussedTagName = function () {
	return this.getActiveElement().getTagName();
};
```

I am wondering how the tests are passing on your side, cause ```getTagName()``` should return ```input``` IMHO which can never be equal to ```new-todo```. I just tried backbone, which fails too for me.

Not sure if we get off-topic now, but other issues I found:

0. ```assertFocussedElementId('new-todo')``` does not take care of new class approach at all - besides the broken ```getTagName()``` thingy.
0. Is the name for ```idSelectors``` in ```page.js``` correct? I guess it is supposed to mean the opposite. Simply [re-check the evaluations of its state](https://github.com/tastejs/todomvc/blob/master/tests/page.js#L19). Besides that it's confusing to read when inverting ```idSelectors``` every time.

I finally made todomvc-aurelia passing the acceptance tests. I'll prepare a separate PR with fixes and improvements in the [todomvc/tests](https://github.com/tastejs/todomvc/tree/master/tests) folders. Or?

mhoyer commented:
Some further issues I see with the current tests. There is rarely correct use of promises and async handling. In fact, no test is invoked with callback function. 

BTW, why was #976 closed so abrupt? I think the idea of having gherkin specs as test base sounds perfect for TodoMVC.

samccone commented:
You should be able to rebase now and get your jscs fix @mhoyer :)

As for your other comments, let me investigate further, thanks for raising these points

samccone commented:
@mhoyer fixed that regression in the tests, that was 100% my bad, sorry
https://github.com/tastejs/todomvc/commit/08ceb647d9581c6db99ddd0a31d0ec39fb462d9e

Naylor55 commented:
what?






At 2015-12-30 00:12:24, "Marcel Hoyer" <notifications@github.com> wrote:


Some further issues I see with the current tests. There is rarely correct use of promises and async handling. In fact, no test is invoked with callback function.

BTW, why was #976 closed so abrupt? I think the idea of having gherkin specs as test base sounds perfect for TodoMVC.

—
Reply to this email directly or view it on GitHub.

mhoyer commented:
@samccone I did a rework of the test suite to fully support promises and async flow. If you are interested check out [my commits on the ```async-tests``` branch in my fork](https://github.com/mhoyer/todomvc/commits/async-tests). I can also provide a PR if you want.

samccone commented:
oh wow @mhoyer that is awesome work, I have some comments but lets move that over to another PR.

For this PR would you mind just dropping all the test work, rebase of of master and then we can land this?

Thanks so much for pushing through this upgrade process with me, I really appreciate your persistence!

mhoyer commented:
> For this PR would you mind just dropping all the test work, rebase of of master and then we can land this?

So I did with #1571. Let me know about your comments on this.

samccone commented:
will do, @mhoyer if you would like to drop those test commits from this PR, we can land it independently :)

mhoyer commented:
Sure I'll drop the ```async-tests``` commits from this PR. But first I need to have #1571 merged into master. At time of writing, Aurelia TodoMVC is not passing the current test suite from ```master``` branch.

samccone commented:
hey @mhoyer mind a rebase so we can land this?

Thanks!

mhoyer commented:
@samccone finally it seems to happen. Just did a last ```jspm update && npm up``` in my [mhoyer/todomvc-aurelia](mhoyer/todomvc-aurelia) and rebased outcome to todomvc master.

:rocket:

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	load() {
+		var storageContent = this.storage.getItem(STORAGE_NAME);
```
const?

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	save() {
+		var simpleItems = _.map(this.items, item => { return {
```
const?

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+		var storageContent = this.storage.getItem(STORAGE_NAME);
+		if (storageContent == undefined) { return; }
+
+		var simpleItems = JSON.parse(storageContent);
```
const?

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	areAllCheckedChanged() {
+		_.each(this.items, i => i.isCompleted = this.areAllChecked);
```
a mutation inside of an each is a bit of code smell.. .can we use a map and an assignment?

passy replied:
:+1: :)

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+		this.save();
+	}
+
+	areAllCheckedChanged() {
```
I am not sure what this method name means? can you explain?

mhoyer replied:
I renamed it to ```onToggleAllChanged```, that is the bound handler for the toggle-all handler.

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	onTitleChanged(todoItem) {
+		if (todoItem.title == '') {
```
triple equal here would be nice

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+		if (title == undefined) { return; }
+
+		title = title.trim();
+		if (title.length == 0) { return; }
```
triple equal

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	onKeyUp(ev) {
+		if (ev.keyCode == ENTER_KEY) {
```
trip equal

samccone commented on examples/aurelia/todo-item.js:
```diff
@@ -0,0 +1,31 @@
+var ENTER_KEY = 13;
```
const

samccone commented on examples/aurelia/todo-item.js:
```diff
@@ -0,0 +1,31 @@
+var ENTER_KEY = 13;
+var ESC_KEY = 27;
```
const

samccone commented on examples/aurelia/todo-item.js:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	onKeyUp(ev) {
+		if (ev.keyCode == ENTER_KEY) {
```
triple equal

samccone commented on examples/aurelia/todo-item.js:
```diff
@@ -0,0 +1,31 @@
+		if (ev.keyCode == ENTER_KEY) {
+			return this.finishEditing();
+		}
+		if (ev.keyCode == ESC_KEY) {
```
triple equal

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+import _ from 'underscore';
+
+var STORAGE_NAME = 'todomvc-aurelia';
+var ENTER_KEY = 13;
```
const

samccone commented on examples/aurelia/todos.js:
```diff
@@ -0,0 +1,140 @@
+import {TodoItem} from './todo-item';
+import _ from 'underscore';
+
+var STORAGE_NAME = 'todomvc-aurelia';
```
const

mhoyer commented:
@samccone, thank you for your feedback. Will have a look into it on weekend.

mhoyer commented:
@samccone I applied your feedback and rebased to current HEAD of master. Thanks again for the recommendations.

mhoyer commented:
@samccone do you need any further support for this PR?

mttmccb commented:
Any ideas what's wrong with this, was just about to do the same as I noticed it doesn't appear to work in firefox etc.... according to #1598

dmethvin commented:
@brykov has [determined the reason](https://github.com/tastejs/todomvc/issues/1598#issuecomment-190190670) so if you can find why the file is using a backslash in the path you will have it solved. :smile:

mhoyer commented:
This PR contains the fix for this issue (see https://github.com/tastejs/todomvc/issues/1492 >> https://github.com/mhoyer/todomvc-aurelia/issues/10). I am just waiting for the PR to be merged. :-(

samccone commented:
Thanks @mhoyer :clap:
